### PR TITLE
Additional stopping functionality

### PR DESCRIPTION
**Pull request by henrymoss**
I have added in an convergence based stopping criterion. We can now tell the loop to stop if two consecutive queries are closer (in euclidean norm) than a user-specified value.

I have also added functionality to use multiple stopping criterion, e.g.specifying  a convergence and max number of steps for the optimization. The loop will stop when any criterion is triggered.

This is demonstrated in Emukit-tutorial-basic-use-of-the-library.ipynb .

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: emukit/core/loop/outer_loop.py [hunk 1]
```diff
@@ -54,18 +54,22 @@ def __init__(self, candidate_point_calculator: CandidatePointCalculator,
...
-                 context: dict=None) -> None:
+    def run_loop(self, user_function: Union[UserFunction, Callable],
+                 stopping_condition: Union[StoppingCondition, int, List[StoppingCondition]],
+                 context: dict = None) -> None:
         """
         :param user_function: The function that we are emulating
-        :param stopping_condition: If integer - a number of iterations to run, if object - a stopping condition object
-                                   that decides whether we should stop collecting more points
+        :param stopping_condition: If integer - a number of iterations to run, if object (or list of such objects) - a stopping
+                        condition object that decides whether we should stop collecting more points
```
Comment (apaleyes) [inline emukit/core/loop/outer_loop.py]: This comment should explain how the list is handled (the fact that we use "or" and not say "and"). Don't be afraid to make it a huge multi-line thing

File: emukit/core/loop/outer_loop.py [hunk 2]
```diff
@@ -54,18 +54,22 @@ def __init__(self, candidate_point_calculator: CandidatePointCalculator,
...
         :param context: The context is used to force certain parameters of the inputs to the function of interest to
                         have a given value. It is a dictionary whose keys are the parameter names to fix and the values
                         are the values to fix the parameters to.
         """
-        if not (isinstance(stopping_condition, int) or isinstance(stopping_condition, StoppingCondition)):
-            raise ValueError("Expected stopping_condition to be an int or a StoppingCondition instance, "
+
+        if not ((isinstance(stopping_condition, list) and
+                all([isinstance(condition, StoppingCondition) for condition in stopping_condition])) or
+                isinstance(stopping_condition, int) or isinstance(stopping_condition, StoppingCondition)):
```
Comment (apaleyes) [inline emukit/core/loop/outer_loop.py]: This `if` is very hard to parse now! Can you break it down somehow? One idea is to create boolean variables for each condition. Something like that:
```
   is_int = isinstance(stopping_condition, int)
   is_single_condition = ...
   is_condition_list = ...
```
Don't have to be like that, the point is to make this `if` readable.

File: emukit/core/loop/stopping_conditions.py [hunk 3]
```diff
@@ -22,7 +27,8 @@ class FixedIterationsStoppingCondition(StoppingCondition):
     """ Stops after a fixed number of iterations """
     def __init__(self, i_max: int) -> None:
         """
-        :param i_max: Maximum number of function observations within the loop, excluding initial points
+        :param i_max: Maximum number of function
+        observations within the loop, excluding initial points
```
Comment (apaleyes) [inline emukit/core/loop/stopping_conditions.py]:
```suggestion
                observations within the loop, excluding initial points
```

normally you want some additional indent for multiline param docstrings like that

File: emukit/core/loop/stopping_conditions.py [hunk 4]
```diff
@@ -31,4 +37,31 @@ def should_stop(self, loop_state: LoopState) -> bool:
         :param loop_state: Object that contains current state of the loop
         :return: True if maximum number of iterations has been reached
         """
-        return loop_state.iteration >= self.i_max
+        status = loop_state.iteration >= self.i_max
+        if status is True:
```
Comment (apaleyes) [inline emukit/core/loop/stopping_conditions.py]: You can use a more informative name, like `should_stop`, and just do `if should_stop`, which reads nicer than `is status is True`

File: emukit/core/loop/stopping_conditions.py [hunk 5]
```diff
@@ -31,4 +37,31 @@ def should_stop(self, loop_state: LoopState) -> bool:
...
+            _log.info("Stopped after {} Evaluations".format(self.i_max))
```
Comment (apaleyes) [inline emukit/core/loop/stopping_conditions.py]:
```suggestion
            _log.info("Stopped after {} evaluations".format(self.i_max))
```

File: emukit/core/loop/stopping_conditions.py [hunk 6]
```diff
@@ -31,4 +37,31 @@ def should_stop(self, loop_state: LoopState) -> bool:
...
+        return status
+
+
+class ConvergenceStoppingCondition(StoppingCondition):
+    """ Stops once we choose a point within eps of a previous
+    point (with respect to euclidean norm), a notion of convergence"""
```
Comment (apaleyes) [inline emukit/core/loop/stopping_conditions.py]: Feels like this comment isn't finished

File: emukit/core/loop/stopping_conditions.py [hunk 7]
```diff
@@ -31,4 +37,31 @@ def should_stop(self, loop_state: LoopState) -> bool:
...
+    def __init__(self, eps: float) -> None:
+        """
+        :param eps: minimum distance between
+        two consecutive x's to keep running the model
```
Comment (apaleyes) [inline emukit/core/loop/stopping_conditions.py]: ditto

File: emukit/core/loop/stopping_conditions.py [hunk 8]
```diff
@@ -31,4 +37,31 @@ def should_stop(self, loop_state: LoopState) -> bool:
...
+
+    def should_stop(self, loop_state: LoopState) -> bool:
+        """
+        :param loop_state: Object that contains current state of the loop
+        :return: True if consecutive x's are too close
+        """
+        if loop_state.iteration < 2:
+            # less than 2 evaluations so cannot calculate distance
+            return False
+        status = math.sqrt(sum((loop_state.X[-1, :] - loop_state.X[-2, :]) ** 2)) <= self.eps
```
Comment (apaleyes) [inline emukit/core/loop/stopping_conditions.py]: anything wrong with `numpy.linalg.norm`? also, same point about boolean variable name

File: notebooks/Emukit-tutorial-basic-use-of-the-library.ipynb [hunk 9]
```diff
@@ -66,9 +66,27 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": 3,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "ename": "ImportError",
```
Comment (apaleyes) [inline notebooks/Emukit-tutorial-basic-use-of-the-library.ipynb]: That's not nice! Can you rerun this notebook to get rid of errors (there is one more below)

File: tests/emukit/core/test_loop_steps.py [hunk 10]
```diff
@@ -22,6 +25,50 @@ def test_fixed_iteration_stopping_condition():
...
+    loop_state_mock = mock.create_autospec(LoopState)
+    loop_state_mock.iteration = 3
+    loop_state_mock.X.return_value(np.array([[0], [1], [1.01]]))
+    assert(any([condition.should_stop(loop_state_mock)for condition in stopping_condition]) is True)
+
+    # check if we stop when we should due to criterion 2
+    loop_state_mock = mock.create_autospec(LoopState)
+    loop_state_mock.iteration = 6
+    loop_state_mock.X.return_value(np.array([[0], [10], [20], [30], [40], [50]]))
+    assert(any([condition.should_stop(loop_state_mock)for condition in stopping_condition]) is True)
```
Comment (apaleyes) [inline tests/emukit/core/test_loop_steps.py]:
```suggestion
    assert(any([condition.should_stop(loop_state_mock) for condition in stopping_condition]) is True)
```

File: tests/emukit/core/test_loop_steps.py [hunk 11]
```diff
@@ -22,6 +25,50 @@ def test_fixed_iteration_stopping_condition():
...
+    assert(stopping_condition.should_stop(loop_state_mock) is False)
+
+    # check if we stop when we should
+    loop_state_mock = mock.create_autospec(LoopState)
+    loop_state_mock.iteration = 5
+    loop_state_mock.X.return_value(np.array([[0], [1], [2], [3], [3.01]]))
+    assert(stopping_condition.should_stop(loop_state_mock) is True)
+
+
+def test_list_of_stopping_conditions():
```
Comment (apaleyes) [inline tests/emukit/core/test_loop_steps.py]: for that test what should ideally be mocked is stopping condition itself. that is, you would have two mock conditions in a list, and manipulate their "should_stop" outputs to test various combinations (none fires, one fires, both fire)

File: emukit/core/loop/stopping_conditions.py (hunk 8 above)
Comment (henrymoss) [reply emukit/core/loop/stopping_conditions.py]: The problem with using np here, is that you end up with a np boolean object. This doesn't work with the assert statements in the tests.

I kept getting assert error False is False
 as one False was the numpy bool and one was the standard python bool

File: emukit/core/loop/outer_loop.py (hunk 2 above)
Comment (mmahsereci) [reply emukit/core/loop/outer_loop.py]: @apaleyes might it we worth doing a "OR" and "AND" stopping condition that handles multiple stopping conditions. So you could have multiple stopping conditions that go into the OrStoppingCondition for example. Then it would be easier to handle the inputs here.

Comment (apaleyes) [reply emukit/core/loop/outer_loop.py]: @mmahsereci in the future maybe, and there might be some nice ways to implement that. but for now that seems overkill. let's deal with it once someone actually needs it

File: emukit/core/loop/stopping_conditions.py (hunk 8 above)
Comment (apaleyes) [reply emukit/core/loop/stopping_conditions.py]: not sure i am following. `linalg.norm` can certainly return a single float, and `float <= float` returns python boolean. what am i missing?

File: emukit/core/loop/outer_loop.py (hunk 2 above)
Comment (apaleyes) [reply emukit/core/loop/outer_loop.py]: after offline chat with @mmahsereci we've decided to cut an issue to do something more flexible here: #273 . @henrymoss please ignore it for now, it is certainly outside the scope of this PR

File: tests/emukit/core/test_loop_steps.py (hunk 11 above)
Comment (apaleyes) [reply tests/emukit/core/test_loop_steps.py]: btw, here is how you can make `should_stop` return different outputs in sequence: https://stackoverflow.com/questions/24897145/python-mock-multiple-return-values

File: emukit/core/loop/stopping_conditions.py (hunk 8 above)
Comment (henrymoss) [reply emukit/core/loop/stopping_conditions.py]: np.linalg.norm returns a numpy float. So the comparison provides a numpy bool

Comment (apaleyes) [reply emukit/core/loop/stopping_conditions.py]: oh, ok, missed that, thanks. in that case how about just converting numpy float to python float?

**Comment by henrymoss** [comment]
Made all the style changes.

**Comment by henrymoss** [comment]
Thanks for that fix. I ran the notebook in a venv and it messed up the init. 

@apaleyes I have a question about using the batch (with local penalization) for a multi-fidelity optimization. It runs fine, but it is hard to diagnose if it actually works (i.e actually performing local penalization across all the fidelity space as well as the search space). Did you ever run any tests on using batch with multi-fidlelity models?

**Comment by apaleyes** [comment]
@henrymoss i don't recall @marpulli or anyone else doing that, tbh

**Comment by henrymoss** [comment]
Yeah so it works. But I'm not sure if it is just a fluke. I will investigate and get back to you!

**Comment by apaleyes** [comment]
@henrymoss thanks! but when you do, let's open a new issue for it
